presto-parquet/src/main/java/com/facebook/presto/parquet/reader/ParquetReader.java

```diff
@@ -408,4 +515,77 @@ private static Block rewriteLongArrayBlock(LongArrayBlock longArrayBlock, Type t
 
         return newBlockBuilder.build();
     }
+
+    private static <T> List<T> listWithNulls(int size)
+    {
+        return Stream.generate(() -> (T) null).limit(size).collect(Collectors.toCollection(ArrayList<T>::new));
+    }
+
+    private RowRanges getRowRanges(int blockIndex)
+    {
+        assert filter != null;
+
+        RowRanges rowRanges = blockRowRanges.get(blockIndex);
+        if (rowRanges == null) {
+            rowRanges = ColumnIndexFilter.calculateRowRanges(FilterCompat.get(filter), blockIndexStores.get(blockIndex),
+                    paths.keySet(), blocks.get(blockIndex).getRowCount());
+            blockRowRanges.set(blockIndex, rowRanges);
+        }
+        return rowRanges;
+    }
+
+    /**
+     * Describes a list of consecutive parts to be read at once. A consecutive part may contain whole column chunks or
+     * only parts of them (some pages).
+     */
+    private class PageRanges
```

Comment:
should we consider merge PageRanges with OffsetRange into one class? They are quite similar

Reply:
It seems the same but we have to convert between int/long.

Reply:
yes, shall we use merge the two classes?

Reply:
The variable 'length' is int in PageRanges but long in OffsetRanges. If we merge, we need to cast, which we should generally avoid casting as much as we can.

Reply:
I am inclined to merge `PageRanges` into `OffsetRange`. We are casting OffsetRange length to int in code below. Doing the merge could remove duplicate code, and save the cast, too.